ca/certificate-authority.go

```diff
@@ -47,6 +49,56 @@ var badSignatureAlgorithms = map[x509.SignatureAlgorithm]bool{
 	x509.ECDSAWithSHA1:             true,
 }
 
+// PKCS#9 attribute for requesting certificate extensions
+var oidExtensionRequest = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 9, 14}
+
+// OID and fixed value for the "must staple" variant of the TLS Feature
+// extension:
+//
+//  Features ::= SEQUENCE OF INTEGER                  [RFC7633]
+//  enum { ... status_request(5) ...} ExtensionType;  [RFC6066]
+//
+// DER Encoding:
+//  30 03 - SEQUENCE (3 octets)
+//  |-- 02 01 - INTEGER (1 octet)
+//  |   |-- 05 - 5
+var oidTLSFeature = asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 24}
+var mustStapleFeatureValue = []byte{0x30, 0x03, 0x02, 0x01, 0x05}
+
+// Extract supported extensions from a CSR.  The following extensions are
+// currently supported:
+//
+// * 1.3.6.1.5.5.7.1.24 - TLS Feature [RFC7633], with the "must staple" value
+//
+// Other requested extensions are silently ignored.
+func extensionsFromCSR(csr *x509.CertificateRequest) (extensions []signer.Extension) {
+	extensions = []signer.Extension{}
+	for _, attr := range csr.Attributes {
+		if !attr.Type.Equal(oidExtensionRequest) {
+			continue
+		}
+
+		for _, extList := range attr.Value {
+			for _, ext := range extList {
+				if ext.Type.Equal(oidTLSFeature) {
+					value, ok := ext.Value.([]byte)
+					if !ok || !bytes.Equal(value, mustStapleFeatureValue) {
+						fmt.Printf("\n\n\n !!! Unsupported value: %v !!! \n\n\n", ext.Value)
```

Comment:
This shouldn't be a `Printf`. Instead, add an error return value to this function and return an error. The error text should be different fora failure to convert the interface type vs the value not matching `mustStapleFeatureValue`.